### PR TITLE
Add #read_multi for fragments

### PR DESCRIPTION
**Pull request by daukadolt**
An MVP version (I'd at least need to add specs) of #read_multi implementation idea

#44

### CODE REVIEW

File: lib/graphql/fragment_cache/schema/lazy_cache_resolver.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            resolved_fragments: {}
+          }
+          @block = block
+
+          @lazy_state[:pending_fragments] << @fragment
+        end
+
+        def resolve
+          unless @lazy_state[:resolved_fragments].key?(@fragment)
+            resolved_fragments = Fragment.read_multi(@lazy_state[:pending_fragments].to_a)
```
Comment (DmitryTsepelev) [inline lib/graphql/fragment_cache/schema/lazy_cache_resolver.rb:23]: You mentioned that you see `write_multi` but also multiple `get` calls. Does it work for you now? If not, can you check what's inside `@lazy_state` in such cases?

**Review by DmitryTsepelev — CHANGES_REQUESTED** [review]
Good start, thank you so much! I've added a couple of questions below. And yes, we need a lot of specs for it 🙂

File: lib/graphql/fragment_cache/schema/lazy_cache_resolver.rb (hunk 1 above)
Comment (daukadolt) [reply lib/graphql/fragment_cache/schema/lazy_cache_resolver.rb:23]: I was going to see whether Redis' `read_multi` was being used, but I think I got carried away and ended up looking into how Cacher does persisting and forgot about `read_multi` 😅.

The behavior I was getting was like this, for some reason I would see first `MGET` with correct cache keys, but weirdly concatenated.
Screen Shot 2021-06-28 at 8 42 44 PM

The problem was that I passed an array of cache keys to `read_multi(*keys)` when I should have splatted the array. I've fixed that in this [commit](https://github.com/DmitryTsepelev/graphql-ruby-fragment_cache/pull/70/commits/c241ef60f1812f54263d31490dda071dcabdde92).

After that, logs look like this:

Screen Shot 2021-06-28 at 8 46 19 PM

plus, the query response now feels significantly faster.

File: lib/graphql/fragment_cache/fragment.rb [hunk 2]
```diff
@@ -8,19 +8,50 @@ module FragmentCache
...
+
+          fragments_available_in_context.merge(fetched_fragments_to_values)
+        end
+      end
+
+      attr_reader :options, :path, :context, :keep_in_context
+
+      def initialize(context, keep_in_context = false, **options)
         @context = context
+        @keep_in_context = keep_in_context
```
[the remark this answers is not in the document]
Comment (daukadolt) [reply lib/graphql/fragment_cache/fragment.rb]: I've rolled back the changes, but [this](https://github.com/DmitryTsepelev/graphql-ruby-fragment_cache/blob/09ec09cd2a7fc8a75905ce2bb8840c436772d89d/spec/graphql/fragment_cache/object_helpers_spec.rb#L718) one test is failing now.

Do all the Cache implementations support `multi_read`? If not, should I leave `keep_in_context`?

[the remark this answers is not in the document]
Comment (DmitryTsepelev) [reply lib/graphql/fragment_cache/fragment.rb]: Good point! When multi_read is not supported — we can fallback to read (like I did for [write](https://github.com/DmitryTsepelev/graphql-ruby-fragment_cache/blob/4e35f6e03863306d526438f4fc84d1b7c730d8d5/lib/graphql/fragment_cache/cacher.rb#L13). And still we can get rid of `keep_in_context` — if someone wants things to be fast, he should use the storage that supports batch operations. We'll make a major release for this breaking change

**Comment by DmitryTsepelev** [comment]
Tested locally and and it seems to work! Let's resolve the issue with specs and it will be good to go

**Comment by DmitryTsepelev** [comment]
@daukadolt Hi, did you have a chance to take a look at it? Anything I can help you with?

**Comment by frostmark** [comment]
Yo guys. Maybe I can help to finish that?

**Comment by frostmark** [comment]
Ah, I got it, this branch just needs to update from the master and all tests will be passed. @DmitryTsepelev Take a look please

**Comment by DmitryTsepelev** [comment]
tried to rebase, but it still gives me

```
  1) #cache_fragment context caching when keep_in_context is true calls #read once
     Failure/Error: expect(GraphQL::FragmentCache.cache_store).to have_received(:read).exactly(4)

       (#<GraphQL::FragmentCache::MemoryStore:0x000000010c413518 @default_expires_in=nil, @storage={"fdcef8645ebc3c51d2b94f656ac93230091fa7ce"=>#<struct GraphQL::FragmentCache::MemoryStore::Entry value={"id"=>"1", "title"=>"object test"}, expires_at=nil>}>).read(*(any args))
           expected: 4 times with any arguments
           received: 6 times with any arguments
     # ./spec/graphql/fragment_cache/object_helpers_spec.rb:726:in `block (4 levels) in <compiled>'
```

**Pull request closed without merging**